File: solutions/cpp/easy/0101_symmetric_tree.cpp
#include <vector>
#include <queue>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
 
class Solution {
private:
    bool isMirror(TreeNode* left, TreeNode* right) {
        if (!left || !right) return left == right;
        return left->val == right->val && isMirror(left->left, right->right) && isMirror(left->right, right->left);
    }
public:
    bool isSymmetric(TreeNode* root) {
        return isMirror(root->left, root->right);
    }
};

File: solutions/python/easy/0101_symmetric_tree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def is_mirror(self, left, right):
        if not left or not right:
            return left == right
        return (left.val == right.val
                and self.is_mirror(left.left, right.right)
                and self.is_mirror(left.right, right.left))

    def isSymmetric(self, root: TreeNode) -> bool:
        return self.is_mirror(root.left, root.right)

